### ISSUE TITLE
[native-image] "Undefined symbols for architecture x86_64" on macOS

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#1817]
I've tried to compile GraalSqueak (using `java.awt` for drawing) with native image and `openjdk-8u232-jvmci-19.3-b04` using [`master at 42180c5`](42180c5d084e5c8f413e90c98280c7f83e1163a8) on macOS Catalina.

Compilation fails after stage `[graalsqueak-svm:8120]        write:     868.80 ms` with the following error:

```
Fatal error: java.lang.RuntimeException: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1
...
Undefined symbols for architecture x86_64:
  "_JVM_DoPrivileged", referenced from:
      _Java_java_security_AccessController_doPrivileged__Ljava_security_PrivilegedAction_2 in libjava.a(AccessController.o)
      _Java_java_security_AccessController_doPrivileged__Ljava_security_PrivilegedExceptionAction_2Ljava_security_AccessControlContext_2 in libjava.a(AccessController.o)
      _Java_java_security_AccessController_doPrivileged__Ljava_security_PrivilegedExceptionAction_2 in libjava.a(AccessController.o)
      _Java_java_security_AccessController_doPrivileged__Ljava_security_PrivilegedAction_2Ljava_security_AccessControlContext_2 in libjava.a(AccessController.o)
  "_JVM_GetInheritedAccessControlContext", referenced from:
      _Java_java_security_AccessController_getInheritedAccessControlContext in libjava.a(AccessController.o)
  "_JVM_GetStackAccessControlContext", referenced from:
      _Java_java_security_AccessController_getStackAccessControlContext in libjava.a(AccessController.o)
  "_Java_java_util_prefs_MacOSXPreferencesFile_addChildToNode", referenced from:
      ___svm_cglobaldata_base in graalsqueak-svm.o
  "_Java_java_util_prefs_MacOSXPreferencesFile_addKeyToNode", referenced from:
      ___svm_cglobaldata_base in graalsqueak-svm.o
  "_Java_java_util_prefs_MacOSXPreferencesFile_addNode", referenced from:
      ___svm_cglobaldata_base in graalsqueak-svm.o
  "_Java_java_util_prefs_MacOSXPreferencesFile_getChildrenForNode", referenced from:
      ___svm_cglobaldata_base in graalsqueak-svm.o
  "_Java_java_util_prefs_MacOSXPreferencesFile_getKeyFromNode", referenced from:
      ___svm_cglobaldata_base in graalsqueak-svm.o
  "_Java_java_util_prefs_MacOSXPreferencesFile_removeKeyFromNode", referenced from:
      ___svm_cglobaldata_base in graalsqueak-svm.o
  "_Java_java_util_prefs_MacOSXPreferencesFile_synchronize", referenced from:
      ___svm_cglobaldata_base in graalsqueak-svm.o
ld: symbol(s) not found for architecture x86_64
```

**Comment by olpaw** [comment]
@fniephaus can you provide me with steps to reproduce?
Also is this OSX specific (i.e. does building work on Linux with that same rev-id) ?

**Comment by fniephaus** [comment]
Sure, if you check out https://github.com/hpi-swa/graalsqueak/commit/2fbe3c936ad5656199032cf78ba27eeac513e446 and run `mx squeak-svm`, you should be able to reproduce this.

**Comment by fniephaus** [comment]
Just tried to build it on Linux and it fails with a similar, but different error:

```
Fatal error: java.lang.RuntimeException: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1
...
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/share/native/java/security/AccessController.c:95: undefined reference to `JVM_GetInheritedAccessControlContext'
```

**Comment by olpaw** [comment]
In my Linux box I don't even get to the linking step:
```
[graalsqueak-svm:30466]    classlist:   5,031.40 ms
[graalsqueak-svm:30466]        (cap):     989.31 ms
[graalsqueak-svm:30466]        setup:   2,102.48 ms
[graalsqueak-svm:30466]   (typeflow):  19,872.88 ms
[graalsqueak-svm:30466]    (objects):  19,000.98 ms
[graalsqueak-svm:30466]   (features):  13,786.66 ms
[graalsqueak-svm:30466]     analysis:  54,262.48 ms
[graalsqueak-svm:30466]     (clinit):     949.80 ms
21227 method(s) included for runtime compilation
Deepest level call tree path (62 calls):
...
[graalsqueak-svm:30466]     universe:   2,223.89 ms
Fatal error: com.oracle.svm.core.util.VMError$HostedError: Number of methods for runtime compilation exceeds the allowed limit: 21227 > 5000
```

**Comment by fniephaus** [comment]
@olpaw yes, had to push https://github.com/hpi-swa/graalsqueak/commit/c046deb449b85350790fcc4139370110a3df5c34 for Linux.

**Comment by olpaw** [comment]
Ah! Missing `@TruffleBounday`

**Comment by olpaw** [comment]
ok. now i can reproduce it. i will look into it tomorrow.

**Comment by fniephaus** [comment]
No rush, just wanted to give it a try before the `19.3` release. Thanks!

**Comment by olpaw** [comment]
Ok. With providing trivial impl. for `JVM_GetInheritedAccessControlContext` I can build the image. Now if I try to run the image I run into
```
> /home/pwoegere/OLabs/git/svm-master/graal/sdk/mxbuild/linux-amd64/graalsqueak-svm.image/graalsqueak-svm '/home/pwoegere/Downloads/Squeak5.2-18229-64bit/Squeak5.2-18229-64bit.image' 
[graalsqueak] Running Squeak/Smalltalk on Substrate VM (Graal-compiled)...
[graalsqueak] Image loaded in 2546ms.
ERROR: java.awt.HeadlessException
org.graalvm.polyglot.PolyglotException: java.awt.HeadlessException
	at java.awt.GraphicsEnvironment.checkHeadless(GraphicsEnvironment.java:204)
```
Did GUI mode work with native-image before or is this new?

**Comment by olpaw** [comment]
I have added 7ff42caaa3f09e528af040bc75e2d9f245f967ff a while ago because of focus stealing issues during image build on OSX. I assume this is what gets us in this this new trouble.

**Comment by olpaw** [comment]
When I add `-Djava.awt.headless=false` to your `mx.graalsqueak/native-image.properties` `Args` I run instead into a buildtime issue.
```
[graalsqueak-svm:25442]     analysis:  34,988.69 ms
Error: Classes that should be initialized at run time got initialized during image building:
 sun.awt.dnd.SunDragSourceContextPeer the class was requested to be initialized at build time (from the command line). To see why sun.awt.dnd.SunDragSourceContextPeer got initialized use -H:+TraceClassInitialization
sun.awt.dnd.SunDropTargetContextPeer the class was requested to be initialized at build time (from the command line). To see why sun.awt.dnd.SunDropTargetContextPeer got initialized use -H:+TraceClassInitialization
sun.awt.X11.XDragSourceContextPeer the class was requested to be initialized at build time (). To see why sun.awt.X11.XDragSourceContextPeer got initialized use -H:+TraceClassInitialization
sun.awt.X11.XDropTargetContextPeer the class was requested to be initialized at build time (). To see why sun.awt.X11.XDropTargetContextPeer got initialized use -H:+TraceClassInitialization
```
Some static initializer runs XDragSourceContextPeer at image build time which contradicts your `--initialize-at-run-time=sun.awt.dnd`

**Comment by fniephaus** [comment]
> Did GUI mode work with native-image before or is this new?

So far, we've used a [SDL2-based substitution](https://github.com/hpi-swa/graalsqueak/blob/ff919dc054c4bc371c18cb767e38f21b787ce37e/src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/aot/SqueakDisplaySubstitutions.java#L80) when compiling with native-image and that worked. I now wanted to see if we can get rid of that code, @timfel mentioned this may be possible.

Have you successfully native-compiled any other `java.awt`-based app yet?

**Comment by olpaw** [comment]
> Have you successfully native-compiled any other java.awt-based app yet?

I have not.

**Comment by fniephaus** [comment]
What's the best way to proceed with this now? I already mentioned this is not urgent, but it'd be great if `java.awt` is supported. I might try to compile a basic `JFrame` example and see if I run in similar issues. Also, how do you debug such build problems and how would something like `_Java_java_util_prefs_MacOSXPreferencesFile_addNode` be implemented?

**Comment by fniephaus** [comment]
[This JFrame Example](https://examples.javacodegeeks.com/java-jframe-example/) seems to work just fine: 
![image](https://user-images.githubusercontent.com/2368856/68475840-7e0e4300-0229-11ea-85db-7d39b4ee073d.png)

**Comment by olpaw** [comment]
The `Undefined symbols` issue you reported on top is a real problem and that needs fixing. I already have a PR for master and I hope we can have it still backported to the release branch (having code paths that go through `jdk/src/share/native/java/security/AccessController.c` is something that has to work and must not cause image build failures). Getting `java.awt` to work (without lots of extra config/substitutions) is unrealistic for 19.3. The reason why your JFrame example above works is that you get a fallback image that runs the application on the JVM ;-) Try building with `--no-fallback` and you will see all the reasons why this is still not working.

**Comment by fniephaus** [comment]
Good catch, @olpaw! I didn't know what *fallback* means here and thought it was working. With `--no-fallback`, I can observe the same problem you described earlier (in https://github.com/oracle/graal/issues/1817#issuecomment-551516886). I'd be cool if `java.awt` can be supported in the future, it's absolutely fine if it won't be in `19.3`. :)

**Comment by olpaw** [comment]
The issue with missing `JVM_{DoPrivileged,GetInheritedAccessControlContext,GetStackAccessControlContext}` during image link time is fixed on master and the 19.3 release branch. c35a3342880c9244036d905476ff929f31e251ff
